File: apollo-api/src/main/java/com/apollographql/apollo/api/internal/ResponseReader.java
package com.apollographql.apollo.api.internal;

import com.apollographql.apollo.api.ResponseField;
import com.apollographql.apollo.api.ScalarType;

import java.util.List;

/*
 * ResponseReader is an abstraction for reading GraphQL fields.
 */
public interface ResponseReader {

  String readString(ResponseField field);

  Integer readInt(ResponseField field);

  Long readLong(ResponseField field);

  Double readDouble(ResponseField field);

  Boolean readBoolean(ResponseField field);

  <T> T readObject(ResponseField field, ObjectReader<T> objectReader);

  <T> T readFragment(ResponseField field, ObjectReader<T> objectReader);

  <T> List<T> readList(ResponseField field, ListReader<T> listReader);

  @SuppressWarnings("TypeParameterUnusedInFormals")
  <T> T readCustomType(ResponseField.CustomTypeField field);

  interface ObjectReader<T> {
    T read(ResponseReader reader);
  }

  interface ListReader<T> {
    T read(ListItemReader reader);
  }

  interface ListItemReader {

    String readString();

    Integer readInt();

    Long readLong();

    Double readDouble();

    Boolean readBoolean();

    @SuppressWarnings("TypeParameterUnusedInFormals")
    <T> T readCustomType(ScalarType scalarType);

    <T> T readObject(ObjectReader<T> objectReader);

    <T> List<T> readList(ListReader<T> listReader);
  }
}

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/internal/ResponseReader.kt
package com.apollographql.apollo.api.internal

import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.CustomScalar

/*
 * ResponseReader is an abstraction for reading GraphQL fields.
 */
interface ResponseReader {

  fun selectField(fields: Array<ResponseField>): Int

  fun readString(field: ResponseField): String?

  fun readInt(field: ResponseField): Int?

  fun readDouble(field: ResponseField): Double?

  fun readBoolean(field: ResponseField): Boolean?

  fun <T : Any> readObject(field: ResponseField, block: (ResponseReader) -> T): T?

  fun <T : Any> readList(field: ResponseField, block: (ListItemReader) -> T): List<T?>?

  fun <T : Any> readCustomScalar(field: ResponseField.CustomScalarField): T?

  interface ObjectReader<T : Any> {
    fun read(reader: ResponseReader): T
  }

  interface ListReader<T : Any> {
    fun read(reader: ListItemReader): T
  }

  interface ListItemReader {

    fun readString(): String

    fun readInt(): Int

    fun readDouble(): Double

    fun readBoolean(): Boolean

    fun <T : Any> readCustomScalar(customScalar: CustomScalar): T

    fun <T : Any> readObject(block: (ResponseReader) -> T): T

    fun <T : Any> readList(block: (ListItemReader) -> T): List<T?>
  }
}
